File: sdks/java/src/main/java/io/licensecc/client/DeviceBoundClient.java
package io.licensecc.client;

import java.lang.ref.Cleaner;
import java.lang.ref.Reference;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.locks.ReentrantLock;
import java.util.Arrays;

/** Opaque native owner. Calls block: run on an application worker thread. Only authorize OK permits work. */
public final class DeviceBoundClient implements AutoCloseable {
    public enum Result {
        OK, INVALID_ARGUMENT, UNSUPPORTED_VERSION, UNSUPPORTED_PLATFORM, BUSY, INVALID_STATE,
        RETRY, CONFLICT, DENIED, EXPIRED, WAITING, CALLBACK_REJECTED, CALLBACK_RECEIVED,
        ENROLLMENT_REQUIRED, ONLINE_REQUIRED, RESUME_REQUIRED, INVALID_RESPONSE, PROVIDER_ERROR,
        STORAGE_ERROR, BROWSER_UNAVAILABLE, CANCELLED, INTERNAL_ERROR
    }
    public enum ProviderResult {
        OK, INVALID_ARGUMENT, UNSUPPORTED_VERSION, BUFFER_TOO_SMALL, PROVIDER_UNAVAILABLE,
        HARDWARE_UNAVAILABLE, ACCESS_DENIED, KEY_NOT_FOUND, KEY_CORRUPT, KEY_LOST,
        UNSUPPORTED_ALGORITHM, SIGN_FAILED, IO_ERROR, BUSY, POLICY_VIOLATION, INTERNAL_ERROR
    }
    public enum CheckpointResult {
        NOT_ATTEMPTED, SAVED, UNCHANGED, MISSING, BUSY, STALE, CONFLICT, INVALID, IO_ERROR,
        MIRROR_PENDING, COMMIT_UNKNOWN, LOADED
    }
    /** Persistence and renewal hints are independent of permission to perform protected work. */
    public record Outcome(Result code, ProviderResult providerResult, CheckpointResult checkpointResult,
            boolean renewalDue, BigInteger effectiveTime) { }
    /** Comparison display only: expiresAt is never a caller-supplied authorization clock. */
    public record EnrollmentView(String comparisonCode, BigInteger expiresAt) { }
    public record PrepareResult(Result code, EnrollmentView view) { }

    private static final Cleaner CLEANER = Cleaner.create();
    private static final Outcome BUSY = new Outcome(Result.BUSY, ProviderResult.OK, CheckpointResult.NOT_ATTEMPTED, false, BigInteger.ZERO);
    private final State state;
    private final Cleaner.Cleanable cleanable;
    // This action must not retain the client. The shared lock excludes close from native calls.
    private static final class State implements Runnable {
        final DeviceBoundApi api;
        final ReentrantLock gate = new ReentrantLock();
        long handle;
        State(DeviceBoundApi api, long handle) { this.api = api; this.handle = handle; }
        boolean enter() { return !gate.isHeldByCurrentThread() && gate.tryLock(); }
        long requireOpen() {
            if (handle == 0) throw new IllegalStateException("Device-bound client is closed");
            return handle;
        }
        @Override public void run() {
            if (gate.isHeldByCurrentThread()) throw new IllegalStateException("Cannot close from an active native call");
            gate.lock();
            try {
                long owned = handle;
                handle = 0;
                if (owned != 0) api.close(owned);
            } finally { gate.unlock(); }
        }
    }
    DeviceBoundClient(DeviceBoundApi api) {
        state = new State(api, 0);
        cleanable = CLEANER.register(this, state);
    }
    // Called only by the library before publication; no allocation after the native handle arrives.
    void adopt(long handle) { state.handle = handle; }
    boolean hasHandle() { return state.handle != 0; }

    public Outcome activate() { return invoke(0); }
    public Outcome renew() { return invoke(1); }
    public Outcome authorize() { return invoke(2); }
    public Outcome saveCheckpoint() { return invoke(3); }
    public Outcome abandonPending() { return invoke(4); }
    private Outcome invoke(int operation) {
        if (!state.enter()) return BUSY;
        try {
            long[] raw = newOutcome(0);
            state.api.invoke(state.requireOpen(), operation, raw);
            return decode(raw, 0);
        } finally { state.gate.unlock(); Reference.reachabilityFence(this); }
    }
    public PrepareResult prepare() {
        if (!state.enter()) return new PrepareResult(Result.BUSY, null);
        try {
            long[] raw = new long[2];
            raw[0] = -1;
            byte[] comparison = new byte[15];
            state.api.prepare(state.requireOpen(), raw, comparison);
            Result code = decodeCode(raw[0]);
            if (code != Result.OK) return new PrepareResult(code, null);
            String text = new String(comparison, 0, 14, StandardCharsets.US_ASCII);
            if (comparison[14] != 0 || !text.matches("[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}")) {
                throw new IllegalStateException("Invalid native comparison code");
            }
            return new PrepareResult(code, new EnrollmentView(text, unsigned(raw[1])));
        } finally { state.gate.unlock(); Reference.reachabilityFence(this); }
    }
    public Result launch() { return simple(0, 0); }
    public Result poll(int waitMilliseconds) {
        if (waitMilliseconds < 0 || waitMilliseconds > 1000) throw new IllegalArgumentException("Poll wait must be 0..1000 ms");
        return simple(1, waitMilliseconds);
    }
    public Result cancel() { return simple(2, 0); }
    private Result simple(int operation, int waitMilliseconds) {
        if (!state.enter()) return Result.BUSY;
        try { return decodeCode(state.api.simple(state.requireOpen(), operation, waitMilliseconds)); }
        finally { state.gate.unlock(); Reference.reachabilityFence(this); }
    }
    static Result decodeCode(long code) {
        if (code == 255) return Result.INTERNAL_ERROR;
        if (code < 0 || code > 20) throw new IllegalStateException("Unknown native result");
        return Result.values()[(int) code];
    }
    static Outcome decode(long[] raw, int offset) {
        if (raw.length != offset + 5 || (raw[offset + 1] != 255 && (raw[offset + 1] < 0 || raw[offset + 1] > 14))
                || raw[offset + 2] < 0 || raw[offset + 2] > 11 || raw[offset + 3] < 0 || raw[offset + 3] > 1) {
            throw new IllegalStateException("Invalid native outcome");
        }
        return new Outcome(decodeCode(raw[offset]), raw[offset + 1] == 255 ? ProviderResult.INTERNAL_ERROR : ProviderResult.values()[(int) raw[offset + 1]],
            CheckpointResult.values()[(int) raw[offset + 2]], raw[offset + 3] == 1, unsigned(raw[offset + 4]));
    }
    static long[] newOutcome(int offset) {
        long[] raw = new long[offset + 5];
        Arrays.fill(raw, offset, raw.length, -1);
        return raw;
    }
    private static BigInteger unsigned(long value) { return new BigInteger(Long.toUnsignedString(value)); }
    /** Waits for admitted calls and closes once. Does not retire, save or delete the device key. */
    @Override public void close() {
        try {
            // Calling run directly also makes every concurrent close wait for admission.
            // Do not consume the Cleanable on a rejected reentrant close.
            state.run();
            cleanable.clean();
        } finally { Reference.reachabilityFence(this); }
    }
}

File: sdks/dotnet/src/Licensecc.Client/DeviceBoundClient.cs
using System;
using System.IO;
using System.Runtime.InteropServices;
using System.Text.RegularExpressions;
using System.Threading;

namespace Licensecc.Client.DeviceBound;

/// <summary>Loads the application-owned bridge. Disposing it prevents new opens; existing clients retain their library pin.</summary>
public sealed unsafe class DeviceBoundLibrary : IDisposable
{
    private readonly INativeApi api;
    private readonly object gate=new();
    private bool disposed;
    public DeviceBoundLibrary(string absoluteDllPath) : this(new NativeApi(absoluteDllPath)) { }
    internal DeviceBoundLibrary(INativeApi api) { this.api=api; }
    public (DeviceBoundClient? Client,Outcome Outcome) OpenEnrollment(Configuration configuration) => Open(configuration,false);
    public (DeviceBoundClient? Client,Outcome Outcome) OpenResume(Configuration configuration) => Open(configuration,true);
    private (DeviceBoundClient?,Outcome) Open(Configuration configuration,bool resume)
    {
        ArgumentNullException.ThrowIfNull(configuration);
        lock(gate)
        {
            ObjectDisposedException.ThrowIf(disposed,this);
            using var pin=api.Pin();
            var options=configuration.Encode(api.Options()); var raw=api.Outcome(); IntPtr handle=IntPtr.Zero;
            try
            {
                var result=DeviceBoundClient.Decode(api.Open(resume,&options,&handle,&raw),raw);
                if(result.Code!=Result.Ok)
                {
                    if(handle!=IntPtr.Zero) throw new InvalidDataException("Failed native open returned a handle.");
                    return (null,result);
                }
                if(handle==IntPtr.Zero) throw new InvalidDataException("Successful native open returned no handle.");
                var owner=new ClientHandle(api,handle); handle=IntPtr.Zero;
                try { return (new DeviceBoundClient(api,owner),result); }
                catch { owner.Dispose(); throw; }
            }
            finally { if(handle!=IntPtr.Zero) api.Close(handle); }
        }
    }
    public void Dispose() { lock(gate) { if(disposed)return; disposed=true; api.Dispose(); } }
}

internal sealed class ClientHandle : SafeHandle
{
    private readonly INativeApi api;
    private readonly IDisposable pin;
    internal ClientHandle(INativeApi api,IntPtr handle) : base(IntPtr.Zero,true)
    { this.api=api; pin=api.Pin(); SetHandle(handle); }
    public override bool IsInvalid => handle==IntPtr.Zero;
    protected override bool ReleaseHandle()
    {
        try { api.Close(handle); return true; }
        finally { pin.Dispose(); }
    }
}

/// <summary>Opaque native owner. Calls are blocking; use an application worker thread. Only Authorize returning Ok permits work.</summary>
public sealed unsafe class DeviceBoundClient : IDisposable
{
    private readonly INativeApi api;
    private readonly ClientHandle handle;
    private readonly object gate=new();
    private bool disposed;
    internal DeviceBoundClient(INativeApi api,ClientHandle handle) { this.api=api; this.handle=handle; }
    public Outcome Activate() => Invoke(Call.Activate);
    public Outcome Renew() => Invoke(Call.Renew);
    public Outcome Authorize() => Invoke(Call.Authorize);
    public Outcome SaveCheckpoint() => Invoke(Call.SaveCheckpoint);
    public Outcome AbandonPending() => Invoke(Call.AbandonPending);
    private Outcome Invoke(Call call)
    {
        if(!Monitor.TryEnter(gate)) return new Outcome(Result.Busy);
        try
        {
            ObjectDisposedException.ThrowIf(disposed,this);
            var raw=api.Outcome();
            return Decode(api.Invoke(call,handle.DangerousGetHandle(),&raw),raw);
        }
        finally { GC.KeepAlive(handle); Monitor.Exit(gate); }
    }
    public (Result Code,EnrollmentView? View) Prepare()
    {
        if(!Monitor.TryEnter(gate)) return (Result.Busy,null);
        try
        {
            ObjectDisposedException.ThrowIf(disposed,this);
            var raw=api.View(); var code=DecodeCode(api.Prepare(handle.DangerousGetHandle(),&raw));
            if(code!=Result.Ok) return (code,null);
            if(raw.Size!=sizeof(Abi.View) || raw.Version!=1 || raw.ComparisonCode[14]!=0)
                throw new InvalidDataException("Invalid native comparison layout.");
            var comparison=Configuration.Utf8.GetString(new ReadOnlySpan<byte>(raw.ComparisonCode,14));
            if(!Regex.IsMatch(comparison,"\\A[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}\\z"))
                throw new InvalidDataException("Invalid native comparison code.");
            return (code,new EnrollmentView(comparison,raw.ExpiresAt));
        }
        finally { GC.KeepAlive(handle); Monitor.Exit(gate); }
    }
    public Result Launch() => Simple(api.Launch);
    public Result Poll(uint waitMilliseconds=0)
    {
        if(waitMilliseconds>1000) throw new ArgumentOutOfRangeException(nameof(waitMilliseconds));
        return Simple(pointer=>api.Poll(pointer,waitMilliseconds));
    }
    public Result Cancel() => Simple(api.Cancel);
    private Result Simple(Func<IntPtr,int> call)
    {
        if(!Monitor.TryEnter(gate)) return Result.Busy;
        try { ObjectDisposedException.ThrowIf(disposed,this); return DecodeCode(call(handle.DangerousGetHandle())); }
        finally { GC.KeepAlive(handle); Monitor.Exit(gate); }
    }
    internal static Result DecodeCode(int code) => Enum.IsDefined(typeof(Result),code) ? (Result)code : throw new InvalidDataException("Unknown native result.");
    internal static Outcome Decode(int code,Abi.Outcome raw)
    {
        if(raw.Size!=sizeof(Abi.Outcome) || raw.Version!=1 || raw.Reserved!=0 || raw.RenewalDue>1 ||
            !Enum.IsDefined(typeof(ProviderResult),(int)raw.ProviderResult) || !Enum.IsDefined(typeof(CheckpointResult),(int)raw.CheckpointResult))
            throw new InvalidDataException("Invalid native outcome.");
        return new Outcome(DecodeCode(code),(ProviderResult)raw.ProviderResult,(CheckpointResult)raw.CheckpointResult,raw.RenewalDue==1,raw.EffectiveTime);
    }
    /// <summary>Waits for an admitted call, then closes once. Does not save, retire or delete the native identity.</summary>
    public void Dispose() { lock(gate) { if(disposed)return; disposed=true; handle.Dispose(); } }
}
